### PR TITLE
Enable Continuous Integration with Travis CI

### PR DESCRIPTION
**Pull request by milleniumbug**
For now it doesn't do much (only builds it), but in the future it could run tests, generate packages and so on.

To enable this you need to connect the GitHub account to the free version of Travis CI on https://travis-ci.org/ - once you register there, you can enable CI for specific repositories.

### CODE REVIEW

**Comment by eidheim** [comment]
Great. Only problem is that Travis CI is running Ubuntu 14 I think, which we might stop supporting in some months in order to move to c++14 and std::regex. Do you know if they have plans to upgrade to Ubuntu 16?

**Comment by eidheim** [comment]
A few thoughts on extending the Travis CI:
* compile with -Werror added to CXX extra argument environment variable
* add `scan-build` before `make` (http://clang-analyzer.llvm.org/scan-build.html) with appropriate parameters and report error if issues are reported (for instance if html pages are created)
* compile with both g++ and clang++ (although without clang++ linking, see #158)

**Comment by milleniumbug** [comment]
The official answer seems to be: ["not this year, try using Docker containers instead"](https://github.com/travis-ci/travis-ci/issues/5821).

**Comment by eidheim** [comment]
I guess docker containers is the way to go then, with Debian testing image or something similarly stable and new.

**Comment by milleniumbug** [comment]
Sorry for the commit spam, but I didn't know a better way of testing the changes on Travis. 

This commit introduces container based setup - there are two containers set up there, with build scripts and dockerfiles put into `ci` directory.

File: ci/ubuntu-16.04.docker [hunk 1]
```diff
@@ -0,0 +1,4 @@
+FROM ubuntu:16.04
+
+RUN apt-get update
+RUN apt-get install -y git cmake make g++ libclang-3.5-dev liblldb-3.5-dev clang-format-3.5 pkg-config libboost-filesystem-dev libboost-regex-dev libgtksourceviewmm-3.0-dev aspell-en libaspell-dev
```
Comment (zalox) [inline ci/ubuntu-16.04.docker]:
```Dockerfile
RUN DEBIAN_FRONTEND=noninteractive apt-get update \
&& apt-get install -y \
      git \
      cmake \
      make \
      g++ \
      libclang-3.5-dev \
      liblldb-3.5-dev \
      clang-format-3.5 \
      pkg-config \
      libboost-filesystem-dev \
      libboost-regex-dev \
      libgtksourceviewmm-3.0-dev \
      aspell-en \
      libaspell-dev \
&& rm -rf /var/lib/apt/lists/*
```

Comment (zalox) [inline ci/ubuntu-16.04.docker]: When doing two run statements with apt-get we risk cache invalidation. The one I posted makes sure that doesn't happen.

**Comment by eidheim** [comment]
Thank you! This I think will form a good basis. I'll have a look at the scripts in the coming days, but we can probably merge this soon anyway. @zalox you should probably have a look at what should be done at the Travis CI side, and maybe have a look at these scripts too.

I have since this PR created automatic docker builds for ubuntu 14, debian stable, debian testing, debian unstable and archlinux that I now test all changes on before merge. Testing on various Linux platforms at least have become much easier after learning this. It's even possible to test the GUI with `docker exec debian_testing juci` for instance after a build (although this can't be done on Travis CI of course).

File: ci/ubuntu-16.04.docker [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+RUN DEBIAN_FRONTEND=noninteractive apt-get update \
+&& apt-get install -y \
+      git \
+      cmake \
+      make \
+      g++ \
+      libclang-3.5-dev \
+      liblldb-3.5-dev \
+      clang-format-3.5 \
```
Comment (eidheim) [inline ci/ubuntu-16.04.docker]: Can use the libclang/liblldb/clang-format 3.6 version instead. Currently this is the default llvm version in debian testing/unstable.

File: ci/fedora-23.build [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+cd /home &&
+mkdir build &&
+cd build &&
+cmake -DCMAKE_CXX_COMPILER=g++ .. &&
```
Comment (eidheim) [inline ci/fedora-23.build]: Both .build files: to exit with error code if any warnings you can replace this line with `cmake -DCMAKE_CXX_FLAGS="-Werror" -DCMAKE_CXX_COMPILER=g++ .. &&` if I'm not mistaken.

**Comment by eidheim** [comment]
@zalox I leave merging of this PR up to you, since you might want to study the Travis CI server side.

**Comment by zalox** [comment]
I have some suggestions. I'm trying a docker-compose approach, that makes it very easy to test ci locally.

**Pull request closed without merging**